### PR TITLE
feat: implements `irn_batchFetchMessages`

### PR DESCRIPTION
**Pull request by ganchoradkov**
## Description
Implemented `irn_batchFetchMessages` that is called before subscribing to topics so we can sort all pending messages from the mailbox and avoid out of sync messages and improve consistency in handling when a peer comes online and receives bunch of mailbox messages

## Type of change

- [ ] Chore (non-breaking change that addresses non-functional tasks, maintenance, or code quality improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Draft PR (breaking/non-breaking change which needs more work for having a proper functionality _[Mark this PR as ready to review only when completely ready]_)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?
this test covers exactly this scenario https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/web3wallet/test/sign.spec.ts#L142
## Checklist

- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

# Additional Information (Optional)

Please include any additional information that may be useful for the reviewer.

### CODE REVIEW

File: packages/core/src/controllers/subscriber.ts [hunk 1]
```diff
@@ -263,6 +265,29 @@ export class Subscriber extends ISubscriber {
     }
   }
 
+  private async rpcBatchFetchMessages(subscriptions: SubscriberTypes.Params[]) {
+    if (!subscriptions.length) return;
+    // const relay = subscriptions[0].relay;
```
Comment (bkrem) [inline packages/core/src/controllers/subscriber.ts]: Needed?

**Comment by ganchoradkov** [comment]
waiting on https://github.com/WalletConnect/walletconnect-utils/pull/170 to be merged/release so I can bump versions in this PR

File: packages/core/src/controllers/subscriber.ts [hunk 2]
```diff
@@ -263,6 +265,29 @@ export class Subscriber extends ISubscriber {
...
+    this.logger.debug(`Outgoing Relay Payload`);
+    this.logger.trace({ type: "payload", direction: "outgoing", request });
+    try {
+      const fetchMessages = await createExpiringPromise(
+        this.relayer.request(request).catch((e) => this.logger.warn(e)),
+        this.subscribeTimeout,
+      );
+      return await fetchMessages;
+    } catch (err) {
+      this.relayer.events.emit(RELAYER_EVENTS.connection_stalled);
```
Comment (devceline) [inline packages/core/src/controllers/subscriber.ts:290]: Is this the only possible cause for the error?

Comment (ganchoradkov) [reply packages/core/src/controllers/subscriber.ts:290]: pretty much yes, other failures e.g. in `.relayer.request` are logged

**Comment by bkrem** [comment]
> waiting on [WalletConnect/walletconnect-utils#170](https://github.com/WalletConnect/walletconnect-utils/pull/170) to be merged/release so I can bump versions in this PR

Approved ✅

File: packages/core/src/controllers/relayer.ts [hunk 3]
```diff
@@ -334,6 +334,23 @@ export class Relayer extends IRelayer {
...
   }
 
+  public async handleBatchMessageEvents(messages: RelayerTypes.MessageEvent[]) {
+    if (messages?.length === 0) {
+      this.logger.trace("Batch message events is empty. Ignoring...");
+      return;
+    }
+    try {
+      const sortedMessages = messages.sort((a, b) => a.publishedAt - b.publishedAt);
+      this.logger.trace(`Batch message events sorted`);
```
Comment (bkrem) [inline packages/core/src/controllers/relayer.ts]: Would be good to trace the batch size too:

```suggestion
      this.logger.trace(`Batch of ${sortedMessages.length} message events sorted`);
```

File: packages/core/src/controllers/relayer.ts [hunk 4]
```diff
@@ -334,6 +334,23 @@ export class Relayer extends IRelayer {
...
+      for (const message of sortedMessages) {
+        await this.onMessageEvent(message);
+      }
+      this.logger.trace(`Batch message events processed`);
```
Comment (bkrem) [inline packages/core/src/controllers/relayer.ts]:
```suggestion
      this.logger.trace(`Batch of ${sortedMessages.length} message events processed`);
```

File: packages/core/src/controllers/subscriber.ts [hunk 5]
```diff
@@ -364,6 +389,7 @@ export class Subscriber extends ISubscriber {
       const batches = Math.ceil(this.cached.length / this.batchSubscribeTopicsLimit);
```
Comment (bkrem) [inline packages/core/src/controllers/subscriber.ts]: Nit: `numOfBatches` would be clearer here to indicate that this is a number and not related the `batch` assignment below
```suggestion
      const numOfBatches = Math.ceil(this.cached.length / this.batchSubscribeTopicsLimit);
```

File: packages/core/src/controllers/subscriber.ts [hunk 6]
```diff
@@ -399,6 +425,17 @@ export class Subscriber extends ISubscriber {
     this.onBatchSubscribe(result.map((id, i) => ({ ...subscriptions[i], id })));
   }
 
+  private async batchFetchMessages(subscriptions: SubscriberTypes.Params[]) {
+    if (!subscriptions.length) return;
```
Comment (bkrem) [inline packages/core/src/controllers/subscriber.ts:433]: Is this `subscriptions` guaranteed to always be an array?

File: packages/core/src/controllers/subscriber.ts [hunk 7]
```diff
@@ -399,6 +425,17 @@ export class Subscriber extends ISubscriber {
...
+    this.logger.trace(`Fetching batch messages for ${subscriptions.length} subscriptions`);
+    const response = (await this.rpcBatchFetchMessages(subscriptions)) as {
```
Comment (bkrem) [inline packages/core/src/controllers/subscriber.ts]: Why are we casting? Can we attach the expected return type to `rpcBatchFetchMessages` directly?

File: packages/core/src/controllers/relayer.ts [hunk 8]
```diff
@@ -334,6 +334,23 @@ export class Relayer extends IRelayer {
...
+  public async handleBatchMessageEvents(messages: RelayerTypes.MessageEvent[]) {
+    if (messages?.length === 0) {
+      this.logger.trace("Batch message events is empty. Ignoring...");
+      return;
+    }
+    try {
+      const sortedMessages = messages.sort((a, b) => a.publishedAt - b.publishedAt);
+      this.logger.trace(`Batch message events sorted`);
+      for (const message of sortedMessages) {
+        await this.onMessageEvent(message);
```
Comment (bkrem) [inline packages/core/src/controllers/relayer.ts:346]: Thought: if there's an invalid message event half way through, the throw here will then discard all other (possibly) valid messages afterwards.

Could be safer to try/catch on the individual message level instead?

File: packages/core/src/controllers/subscriber.ts (hunk 6 above)
Comment (ganchoradkov) [reply packages/core/src/controllers/subscriber.ts:429]: yes, when there are no messages, response is `[]`

Comment (ganchoradkov) [reply packages/core/src/controllers/subscriber.ts:429]: I can still check that value exist as a precaution

Comment (ganchoradkov) [reply packages/core/src/controllers/subscriber.ts:433]: Yes, it uses same values/types as `batchSubscribe` which has been in prod for many months

**Review by devceline — APPROVED** [review]
LGTM

File: packages/core/src/controllers/subscriber.ts [hunk 9]
```diff
@@ -277,15 +277,19 @@ export class Subscriber extends ISubscriber {
     };
     this.logger.debug(`Outgoing Relay Payload`);
     this.logger.trace({ type: "payload", direction: "outgoing", request });
+    let result;
     try {
       const fetchMessages = await createExpiringPromise(
```
Comment (bkrem) [inline packages/core/src/controllers/subscriber.ts]:
```suggestion
      const fetchMessagesPromise = await createExpiringPromise(
```

**Review by bkrem — APPROVED** [review]

**Pull request merged**